dsgrid/config/representative_period_time_dimension_config.py

```diff
@@ -44,10 +48,76 @@ def check_dataset_time_consistency(self, load_data_df):
-        # Account for time zone.
+    def get_time_dataframe(self):
+        time_cols = self.get_timestamp_load_data_columns()
+        schema = StructType(
+            [StructField(time_col, IntegerType(), False) for time_col in time_cols]
+        )
+
+        model_time = self.list_expected_dataset_timestamps()
+        df_time = _get_spark_session().createDataFrame(model_time, schema=schema)
+
+        return df_time
+
+    def get_time_dataframe_in_model_timezone(self):
+        return self.get_time_dataframe()
+
+    def convert_dataframe(self, df=None, project_time_dim=None, df_meta=None):
+        """convert df time column(s) based on project_time_dim.
+        df_meta is load_data_lookup with time_zone column
+        in pyspark: 1=Sunday, 7=Saturday
+        in pandas: 0:Monday, 6:Sunday
+        the mapping is: day_of_week = [(i+7-2)%7 for i in pyspark.dayofweek]
+        """
+        if project_time_dim is None:
+            return df
+        if (
+            project_time_dim.list_expected_dataset_timestamps()
+            == self.list_expected_dataset_timestamps()
+        ):
+            return df
+
+        time_cols = self.get_timestamp_load_data_columns()
+        ptime_col = project_time_dim.get_timestamp_load_data_columns()
+        assert len(ptime_col) == 1, ptime_col
+        ptime_col = ptime_col[0]
+
+        # get unique timezones
+        if df_meta is not None:
+            df = df.join(df_meta.select("id", "time_zone"), on="id")
+        geo_tz_values = [row.time_zone for row in df.select("time_zone").distinct().collect()]
+        geo_tz_names = [get_timezone(tz).tz_name for tz in geo_tz_values]
+
+        # create time map
+        project_time_df = project_time_dim.get_time_dataframe()
+        session_tz = _get_spark_session().conf.get("spark.sql.session.timeZone")
+        idx = 0
+        for tz_value, tz_name in zip(geo_tz_values, geo_tz_names):
+            local_time_df = project_time_df.withColumn("time_zone", F.lit(tz_value)).withColumn(
+                "local_time",
+                F.from_utc_timestamp(F.to_utc_timestamp(F.col(ptime_col), session_tz), tz_name),
+            )
+            select = [ptime_col, "time_zone"]
+            for col in time_cols:
+                func = col.replace("_", "")
+                expr = f"{func}(local_time) AS {col}"
+                if col == "day_of_week":
+                    expr = f"mod(dayofweek(local_time)+7-2, 7) AS {col}"
+                select.append(expr)
+            local_time_df = local_time_df.selectExpr(*select)
+            if idx == 0:
+                time_df = local_time_df
+            else:
+                time_df = time_df.union(local_time_df)
+            idx += 1
+
+        # join all
+        join_keys = time_cols + ["time_zone"]
+        select = [
+            col if col not in time_cols else F.col(col).cast(IntegerType()) for col in df.columns
+        ]
+        df = df.select(*select).join(time_df, on=join_keys).drop(*join_keys)
```

Comment:
I can't pretend to understand this code just by reading it. I trust daylight savings time is getting handled somewhere?

Reply:
Yes, spark does the daylight savings handling internally, provided that the input timestamps are correct and time_zone-aware and that we stick to spark method when converting between time_zones